Update section on WASM (#25)
hyperledger/iroha-2-docs: src/guide/advanced/intro.md

crate-type = ['cdylib']

[dependencies]
iroha_wasm = { git = "https://github.com/hyperledger/iroha/", branch = "iroha2" }

parity-scale-codec = {version = "2.3.1", default-features = false }
```

> Note that the crate type is `cdylib`. Most Rust code is linked in a non-portable architecture and OS-specific static manner, but WASM is a portable format. Since C ABI is the lingua franca of the programming world and there is no other stable Rust ABI, Iroha relies on the C-linkage to generate WASM bindings. Note that you don't have to worry about C FFI in the smart contract itself, most of that is taken care for you.

Reviewer: ```suggestion
Note that the crate type is `cdylib`. Most Rust code is linked in a non-portable architecture and OS-specific static manner, but WASM is a portable format. Since C ABI is the *lingua franca* of the programming world and there is no other stable Rust ABI (yet), Iroha relies on the C-linkage to generate WASM bindings. Thankfully, you don't need to worry about foreign function interface (FFI)-related problems like `unsafe`, `repr(C)`, padding, alignment etc., `iroha_wasm` takes care of all that for you. 
```

> `iroha_wasm` is the crate that contains all of the bindings, macros and trait implementations that you'd need to write the program, most notably the `iroha_wasm` attribute macro. This crate also exposes `data_model` which contains all of the basic ISI and types. `parity-scale-codec` is (as of today) the chosen serialisation format with a strong possibility of getting replaced by a different(custom) serialization format in the near future, as it seems to dominate the binary size[^2].

Reviewer: `iroha_wasm` attribute macro sounds like it could use a different name. Maybe `smartcontract_entrypoint`, or just `entrypoint`?

Reviewer: ```suggestion
`iroha_wasm` is the crate that contains all of the bindings, macros and trait implementations that you'd need to write the program, most notably the `iroha_wasm` attribute macro. The crate also exposes our `data_model` which contains all of the basic ISI and types. `parity-scale-codec` is (as of today) the chosen serialisation format with a strong possibility of getting replaced with a different (custom) serialisation format in the near future, as it seems to dominate the binary size[^2].
```

Replace **with**. 

`-ise` `-ize` consistency.

> Now that the preliminaries are nailed down, we get to write a basic smart contract. In the `src/lib.rs` you should write the following:

Reviewer: ```suggestion
Now that we have the preliminaries nailed down, we get to write some code for our smart contract. In the `src/lib.rs` you should write the following:
```
The first is a stylistic choice. 

The second is emphasising that so far what we've been doing is a smart contract, just not source code.

#[iroha_wasm::iroha_wasm]
fn smartcontract_entry_point(_account_id: AccountId) {

>     // Find all Registered domains

Reviewer: I think we should either keep this comment, or remove also remove the other.

`cargo run --release` will submit the instruction and run it for you (be sure to have a peer up).

> What this smart contract does is query all of the currently existing domains, put the results into a `std::vec::Vec`, which in this case has to be imported from `alloc`, as we use `no_std` (more on that later), which is then used to add the user named `mad_hatter` to all of the existing domains.

Reviewer: There should always be a space before a parenthesis, regardless of formatting.

> What this smart contract does is query all of the currently existing domains, put the results into a `std::vec::Vec`, which in this case has to be imported from `alloc`, as we use `no_std`(more on that later), which is then used to add the user named `mad_hatter` to all of the existing domains.

Reviewer: Mistake?

codegen-units = 1   # Further reduces binary size but increases compilation time
```

> A few things should be noted; first, Rust stores a lot of debug information(even when compiled in `release` mode) to aid you when you encounter a panic, so that you can debug your application. All this debug information bloats the binary significantly, so instead of storing it, panics are directed to immediately abort the execution. By doing this no crucial functionality is lost, despite to what you might think. Panics are a rare event in the Rust world and the reason of failure to execute a WASM blob isn't logged on the peer, so even if you had some useful information in the panic message, you have no reliable way of retrieving it.

Reviewer: ```suggestion
Rust stores a lot of debug information (even when compiled in `release` mode), which is (as the name suggests)  used for debugging a panic in your Rust application. As you would expect, this information increases the size of the compiled WASM significantly. Normally, this would be a worthwhile trade-off, but not in our case. Firstly, since the WASM is stored on-chain, it will be permanently recorded in some block and take space on every full node of an Iroha network (which by the way, stores a lot of its information in RAM), so storage space for WASM is at a premium. Secondly, once the WASM smart contract is stored on-chain, the debug information is no longer accessible. Indeed, you shouldn't debug on a peer. Instead, you should replicate the conditions that caused the panic locally and debug on your personal machine. 
```

> Another step that was already taken involves working under `no_std` environment. All of our size-related woes stem from Rust being predominantly statically linked. As such, breaking the standard library into more manageable crates, like using `alloc::vec` instead of `std::vec` can help us reduce the size and compilation time[^3].

Reviewer: ```suggestion
Another step that we've already taken involves working under a  `no_std` environment. All of our size-related woes stem from Rust being predominantly statically linked. As such, breaking the standard library into more manageable crates, like using `alloc::vec` instead of `std::vec` can help us reduce the size and compilation time[^3].
```

We, including the reader. _Was_ in this context implies "inside `iroha_wasm`".

> Next, you're advised to re-compile `libcore` and any other standard library crate(e.g. `alloc`) to exclude the leftover panicking infrastructure that comes with prebuilt `core` library(it's also possible `wasm-opt` can be configured to remove these sections for you):

Reviewer: ```suggestion
Next, you're advised to re-compile `libcore` and any other standard library crate (e.g. `alloc`) to exclude the leftover panic-related code that comes with prebuilt the `core` library[^3]:

[^3] `wasm-opt` can also be used to remove the debug sections. 
```

```bash
cargo +nightly build -Z build-std -Z build-std-features=panic_immediate_abort --target wasm32-unknown-unknown
```

Unfortunately, this is an unstable feature. In other words, the developers of the Rust programming language can decide to change how this works, or remove this option entirely.

> Finally, you can use an automated tool to optimise the size of the WASM using [`wasm-opt`](https://github.com/WebAssembly/binaryen), or use [`twiggy`](https://rustwasm.github.io/twiggy/) to guide your manual optimisation efforts. Using `wasm-opt` is highly advised because it will often significantly reduce your binary size, e.g.: `$ wasm-opt -Os -o output.wasm input.wasm`.

Reviewer: ```suggestion
Finally, you can use an automated tool to optimise the size of the WASM using [`wasm-opt`](https://github.com/WebAssembly/binaryen), or use [`twiggy`](https://rustwasm.github.io/twiggy/) to guide your manual optimisation efforts. Using `wasm-opt` is highly advised because it will often significantly reduce your binary size, e.g.: 

```bash
wasm-opt -Os -o output.wasm input.wasm
```

Reviewer: We're not using `$` to signal "without `sudo`" anywhere else in the document.
